pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -55,268 +131,477 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 		return err
 	}
 
-	pred := predicate.Funcs{
-		UpdateFunc: func(e event.UpdateEvent) bool {
-			newObj := e.ObjectNew.(*kubevirtv1alpha1.NodeMaintenance)
-			return !newObj.DeletionTimestamp.IsZero()
-		},
-	}
+	return c.Watch(&source.Kind{Type: &corev1.Node{}}, &handler.EnqueueRequestForObject{})
+}
+
+// Reconcile monitors Nodes and creates the MachineRemediation object when the node has reboot annotation.
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconcile.Result, error) {
 
-	// Create a source for watching noe maintenance events.
-	src := &source.Kind{Type: &kubevirtv1alpha1.NodeMaintenance{}}
+	r.reqLogger = log.WithFields(log.Fields{"Request.Namespace": request.Namespace, "Request.Name": request.Name})
+	r.reqLogger.Debug("Reconciling node", request.Name)
 
-	// Watch for changes to primary resource NodeMaintenance
-	err = c.Watch(src, &handler.EnqueueRequestForObject{}, pred)
+	// Get node from request
+	node := &corev1.Node{}
+
+	if err := r.client.Get(context.TODO(), request.NamespacedName, node); err != nil {
+		if errors.IsNotFound(err) {
+			r.reqLogger.Info("node not found")
+			return reconcile.Result{}, nil
+		}
+		r.reqLogger.Errorf("node not found. name %s error %v", request.Name, err)
+		return reconcile.Result{}, err
+	}
+
+	// parse status & spec, create initial status object on first call
+	specAnnoData, statusAnnoData, node, err := r.parseAnnotations(node)
 	if err != nil {
-		return err
+		r.reqLogger.Errorf("invalid value of annotation. name %s error %v", request.Name, err)
+		return reconcile.Result{}, err
 	}
-	return nil
+
+	if specAnnoData != nil { // maint. annotation exists. user requests to transition toware
+		if ret, res, err := r.handleLease(specAnnoData, statusAnnoData, node); ret {
+			return *res, err
+		}
+
+		res, err := r.processMaintModeOn(statusAnnoData, node)
+		return res, err
+
+	} else if statusAnnoData != nil { // spec deleted but annotation exists -> can be request to exit maintenance mode.
+
+		res, err := r.processMaintModeOff(statusAnnoData, node)
+		return res, err
+	}
+
+	// nothin to do on the node, no relevent annotations have been set
```

Comment:
What if the lease exists and is owned by us?

Reply:
depending on the mode:  if lifecycle.openshift.io/maintenance exists

Reply:
There is a third case: the lease exists, is owned by us, but has expired (or is about to)